// angular-otp-box/esm5/lib/directives/timer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { Subject, Subscription, timer } from 'rxjs';
import { switchMap, take, tap } from 'rxjs/operators';
var CounterDirective = /** @class */ (function () {
    function CounterDirective() {
        var _this = this;
        this._counterSource$ = new Subject();
        this._subscription = Subscription.EMPTY;
        this.interval = 1000;
        this.value = new EventEmitter();
        this._subscription = this._counterSource$.pipe(switchMap((/**
         * @param {?} __0
         * @return {?}
         */
        function (_a) {
            var interval = _a.interval, count = _a.count;
            return timer(0, interval).pipe(take(count), tap((/**
             * @return {?}
             */
            function () { return _this.value.emit(--count); })));
        }))).subscribe();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    CounterDirective.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        this.startTimer();
    };
    /**
     * @return {?}
     */
    CounterDirective.prototype.startTimer = /**
     * @return {?}
     */
    function () {
        this._counterSource$.next({ count: this.counter, interval: this.interval });
    };
    /**
     * @return {?}
     */
    CounterDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this._subscription.unsubscribe();
    };
    CounterDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[counter]'
                },] }
    ];
    /** @nocollapse */
    CounterDirective.ctorParameters = function () { return []; };
    CounterDirective.propDecorators = {
        counter: [{ type: Input }],
        value: [{ type: Output }]
    };
    return CounterDirective;
}());
export { CounterDirective };
if (false) {
    /**
     * @type {?}
     * @private
     */
    CounterDirective.prototype._counterSource$;
    /**
     * @type {?}
     * @private
     */
    CounterDirective.prototype._subscription;
    /** @type {?} */
    CounterDirective.prototype.counter;
    /** @type {?} */
    CounterDirective.prototype.interval;
    /** @type {?} */
    CounterDirective.prototype.value;
}
