// angular-otp-box/esm2015/lib/directives/timer.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Output, EventEmitter } from '@angular/core';
import { Subject, Subscription, timer } from 'rxjs';
import { switchMap, take, tap } from 'rxjs/operators';
export class CounterDirective {
    constructor() {
        this._counterSource$ = new Subject();
        this._subscription = Subscription.EMPTY;
        this.interval = 1000;
        this.value = new EventEmitter();
        this._subscription = this._counterSource$.pipe(switchMap((/**
         * @param {?} __0
         * @return {?}
         */
        ({ interval, count }) => timer(0, interval).pipe(take(count), tap((/**
         * @return {?}
         */
        () => this.value.emit(--count))))))).subscribe();
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        this.startTimer();
    }
    /**
     * @return {?}
     */
    startTimer() {
        this._counterSource$.next({ count: this.counter, interval: this.interval });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this._subscription.unsubscribe();
    }
}
CounterDirective.decorators = [
    { type: Directive, args: [{
                selector: '[counter]'
            },] }
];
/** @nocollapse */
CounterDirective.ctorParameters = () => [];
CounterDirective.propDecorators = {
    counter: [{ type: Input }],
    value: [{ type: Output }]
};
if (false) {
    /**
     * @type {?}
     * @private
     */
    CounterDirective.prototype._counterSource$;
    /**
     * @type {?}
     * @private
     */
    CounterDirective.prototype._subscription;
    /** @type {?} */
    CounterDirective.prototype.counter;
    /** @type {?} */
    CounterDirective.prototype.interval;
    /** @type {?} */
    CounterDirective.prototype.value;
}
